community/cypher/runtime-util/src/main/java/org/neo4j/cypher/internal/runtime/LongArrayHashTable.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2002-2018 "Neo4j,"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.runtime;
+
+import org.neo4j.helpers.collection.Pair;
+
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.CONTINUE_PROBING;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.NOT_IN_USE;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.SLOT_EMPTY;
+import static org.neo4j.cypher.internal.runtime.LongArrayHash.VALUE_FOUND;
+
+/**
+ * This hash table is used as a backing store for the keys of set, map and multi-map where the keys are arrays of longs.
+ */
+class LongArrayHashTable
+{
+    public final long[] keys;
+    public final int width;
+    public final int capacity;
+    private int numberOfEntries;
+    private int resizeLimit;
+
+    private static final double LOAD_FACTOR = 0.75;
+
+    int tableMask;
+
+    LongArrayHashTable( int capacity, int width )
+    {
+        resizeLimit = (int) (capacity * LOAD_FACTOR);
+        tableMask = Integer.highestOneBit( capacity ) - 1;
+        keys = new long[capacity * width];
+        this.width = width;
+        java.util.Arrays.fill( keys, NOT_IN_USE );
+        this.capacity = capacity;
+    }
+
+    /**
+     * Signals whether it's time to size up or not. The actual resizing is done slightly differently depending on if this is a map or set. Maps have, in
+     * addition to a hash table also a separate array for the values.
+     *
+     * @return true if the number of keys in the hash table has reached capacity.
+     */
+    boolean timeToResize()
+    {
+        return numberOfEntries == resizeLimit;
```

Comment:
I have some spider-sense that tingles here. If a user of this table claims twice without checking timeToResize, it will never resize. Putting in a `>=` or some assert might provide better safety for future devs.